src/server/ocpp/services/OCPPService.ts

```diff
@@ -1027,8 +1048,16 @@ export default class OCPPService {
           lastTransaction.stop.totalInactivitySecs + lastTransaction.stop.extraInactivitySecs);
         // Build extra inactivity consumption
         await OCPPUtils.buildExtraConsumptionInactivity(tenantID, lastTransaction);
-        // OCPI: Post the CDR
-        await OCPPUtils.processOCPITransaction(tenantID, lastTransaction, chargingStation, TransactionAction.END);
+        // Indicator needed to decide if transaction is handled with either Gireve or Hubject.
+        if (lastTransaction.ocpiData?.session) {
+          // OCPI: Post the CDR
+          await OCPPUtils.processOCPITransaction(tenantID, lastTransaction, chargingStation, TransactionAction.END);
+        } else if (lastTransaction.oicpData?.session) {
+          // OICP: Post the CDR
+          const user = await UserStorage.getUser(tenantID, lastTransaction.userID);
+          lastTransaction.user = user;
+          await OCPPUtils.processOICPTransaction(tenantID, lastTransaction, chargingStation, TransactionAction.END);
+        }
```

Comment:
We really have to make an abstract class for roaming hub type/protocol ...

Reply:
yes, definitely